### PR TITLE
#162162997 configure jwt authentication

### PR DESCRIPTION
**Pull request by aayebare**
**What does this PR do?**

- Generates a token for every successful user registration
- Returns a token upon successful user login

**Description of the task to be completed?**

- Configure jwt authentication to generate token and return it to the user upon successful user login

**How should this be manually tested?**
- from this repository, clone the branch using ```git clone ft-jwt-authentication-162162997``` 
command in your git bash terminal
- create a virtual environment and use the command ```pip install -r requirements.txt``` to install the application's dependencies in the created virtual environment
- run the application server with ```python manage.py runserver ``` command in the terminal
- add a json object with the following parameters to the ```api/users/``` endpoint using postman
```
{
  "user": {
        "email": "email@email.com,
        "username": "username",
        "password": "password211"
  }
}

```
- use the ```api/users/login``` endpoint to log in and pass in the user email and password in the fomat specified above.
- for the endpoints that require authentication before being accessed, pass a header in the format ```Token <token_credentials>```. The token is generated once the user logs in.

**Relevant background context?**
- Without authentication, any type of users can access any resources of the application. Generating a token helps to authenticate users before they are given access to the requested resources.

**Relevant pivotal tracker stories?**
#162162997

**screenshots**
screen shot 2018-12-04 at 9 48 07 am
screen shot 2018-12-04 at 9 49 50 am

### CODE REVIEW

File: authors/apps/authentication/backends.py [hunk 1]
```diff
@@ -1,10 +1,56 @@
...
-"""Configure JWT Here"""
+from django.conf import settings
 
+from rest_framework import authentication, exceptions
+
+from .models import User
+
+
+class JWTAuthentication(authentication.BaseAuthentication):
+    authentication_header_prefix = 'Token'
```
Comment (arthurarty) [inline authors/apps/authentication/backends.py:11]: @m-asiimwe There is an issue on this line. Please change authentication_header_prefix to 'Bearer and not 'Token'. This is because we will be passing in tokens in the format Bearer <token>.

**Comment by malep2007** [comment]
@m-asiimwe, code coverage has dropped significantly by 4%, please ensure to write more tests so to ensure that your contribution adds value to the work being done by others.

File: authors/apps/authentication/backends.py [hunk 2]
```diff
@@ -1,10 +1,56 @@
...
+        # returns none if no authentication is required for a given request
+        request.user = None
+
+        auth_header = authentication.get_authorization_header(request).split()
+        auth_header_prefix = self.authentication_header_prefix.lower()
+
+        if not auth_header:
+            return None
+
+        if len(auth_header) == 1:
```
Comment (malep2007) [inline authors/apps/authentication/backends.py]: this is ```if statement``` and the next can surely be merged and you can get one return statement

File: authors/apps/authentication/serializers.py [hunk 3]
```diff
@@ -11,19 +11,17 @@ class RegistrationSerializer(serializers.ModelSerializer):
     # Ensure passwords are at least 8 characters long, no longer than 128
     # characters, and can not be read by the client.
     password = serializers.CharField(
-        max_length=128,
-        min_length=8,
-        write_only=True
-    )
+        max_length=128, min_length=8, write_only=True)
```
Comment (malep2007) [inline authors/apps/authentication/serializers.py]: Why did you change the format of the parameters in the ```CharField()```?

**Review by malep2007 — CHANGES_REQUESTED** [review]
Write more tests

**Comment by malep2007** [comment]
The format of your JSON object in the commit message is not correct, can you adjust it from 
```
{"user":
{
"username":"",
"email":"",
"password":""
}
}
```
to the correct format to make it legible.

File: authors/apps/authentication/backends.py (hunk 1 above)
Comment (aayebare) [reply authors/apps/authentication/backends.py:11]: okay.. but is there any particular advantage Bearer gives us over Token?

File: authors/apps/authentication/backends.py (hunk 2 above)
Comment (aayebare) [reply authors/apps/authentication/backends.py]: noted. thanks

File: authors/apps/authentication/serializers.py (hunk 3 above)
Comment (okellogabrielinnocent) [reply authors/apps/authentication/serializers.py]: I believe that could have been pylint

Comment (okellogabrielinnocent) [reply authors/apps/authentication/serializers.py]: Nice update on the work

File: authors/apps/authentication/backends.py (hunk 1 above)
Comment (aayebare) [reply authors/apps/authentication/backends.py:11]: The team is okay with using 'Token' as long as it doesn't affect the functionality of the application which it doesn't

**Review by malep2007 — APPROVED** [review]
